synapse/third_party_rules/access_rules.py

```diff
@@ -0,0 +1,375 @@
+        """
+        new_rule = event.content.get("rule")
+
+        # Check for invalid values.
+        if not self._is_rule_name_valid(new_rule):
+            return False
+
+        # Make sure we don't apply "direct" if the room has more than two members.
+        if new_rule == ACCESS_RULE_DIRECT:
+            member_events_count = 0
+            for key, event in state_events.items():
+                if key[0] == EventTypes.Member:
+                    member_events_count += 1
+
+            if member_events_count > 2:
+                return False
+
+        prev_rules_event = state_events.get((ACCESS_RULES_TYPE, ""))
+
+        # Now that we know the new rule doesn't break the "direct" case, we can allow any
+        # new rule in rooms that had none before.
+        if prev_rules_event is None:
+            return True
+
+        prev_rule = prev_rules_event.content.get("rule")
+
+        # Currently, we can only go from "restricted" to "unrestricted".
+        if prev_rule == ACCESS_RULE_RESTRICTED and new_rule == ACCESS_RULE_UNRESTRICTED:
+            return True
+
+        return False
+
+    def _apply_restricted(self, event):
+        """Implements the checks and behaviour specified for the "restricted" rule.
+
+        Args:
+            event (synapse.events.EventBase): The event to check.
+        Returns:
+            bool, True if the event can be allowed, False otherwise.
+        """
+        # "restricted" currently means that users can only invite users if their server is
+        # included in a limited list of domains.
+        invitee_domain = DomainRuleChecker._get_domain_from_id(event.state_key)
+        return invitee_domain not in self.domains_forbidden_when_restricted
+
+    def _apply_unrestricted(self):
+        """Implements the checks and behaviour specified for the "unrestricted" rule.
+
+        Returns:
+            bool, True if the event can be allowed, False otherwise.
+        """
+        # "unrestricted" currently means that every event is allowed.
+        return True
+
+    def _apply_direct(self, event, state_events):
+        """Implements the checks and behaviour specified for the "direct" rule.
+
+        Args:
+            event (synapse.events.EventBase): The event to check.
+            state_events (dict[tuple[event type, state key], EventBase]): The state of the
+                room before the event was sent.
+        Returns:
+            bool, True if the event can be allowed, False otherwise.
+        """
+        # "direct" currently means that no member is allowed apart from the two initial
+        # members the room was created for (i.e. the room's creator and their first
+        # invitee).
+        if event.type != EventTypes.Member and event.type != EventTypes.ThirdPartyInvite:
+            return True
+
+        # Get the m.room.member and m.room.third_party_invite events from the room's
+        # state.
+        member_events = []
```

Comment:
Given we only use the `state_key` I'd probably make this a set of the `state_keys` and call it something like `existing_members` or something. That way it simplifies things further down since you can write things like `if target in existing_members: ...`